pkg/report/linux_test.go

```diff
@@ -230,3 +235,187 @@ func TestLinuxSymbolizeLine(t *testing.T) {
+				rawBytes:       []byte{0x31, 0xc0, 0xe8, 0xf5, 0xbf, 0xf7, 0xff},
+				decompileFlags: []DecompilerFlag{},
+				offset:         2,
+			},
+		},
+		{
+			input: "31c0 <e8f5> bff5 f7ff ff00",
+			output: parsedOpcodes{
+				rawBytes:       []byte{0x31, 0xc0, 0xe8, 0xf5, 0xbf, 0xf5, 0xf7, 0xff, 0xff, 0x00},
+				decompileFlags: []DecompilerFlag{},
+				offset:         2,
+			},
+		},
+		{
+			input: "<00112233> 44556677",
+			output: parsedOpcodes{
+				rawBytes:       []byte{0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77},
+				decompileFlags: []DecompilerFlag{},
+				offset:         0,
+			},
+		},
+	}
+
+	beReporter := prepareLinuxReporter(targets.S390x, t)
+	runValidTests(beReporter, validBeTests)
+	runInvalidTests(leReporter, invalidTests)
+
+	// Test ARM Thumb detection.
+
+	validArmTests := []validTest{
+		{
+			input: "0011 (2233) 4455",
+			output: parsedOpcodes{
+				rawBytes:       []byte{0x11, 0x00, 0x33, 0x22, 0x55, 0x44},
+				decompileFlags: []DecompilerFlag{FlagForceArmThumbMode},
+				offset:         2,
+			},
+		},
+		{
+			input: "(33221100) 77665544",
+			output: parsedOpcodes{
+				rawBytes:       []byte{0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77},
+				decompileFlags: []DecompilerFlag{},
+				offset:         0,
+			},
+		},
+	}
+
+	armReporter := prepareLinuxReporter(targets.ARM, t)
+	runValidTests(armReporter, validArmTests)
+}
+
+func TestDisassemblyInReports(t *testing.T) {
+	archPath := filepath.Join("testdata", "linux", "decompile")
+	subFolders, err := ioutil.ReadDir(archPath)
+	if err != nil {
+		t.Fatalf("disassembly reports failed: %v", err)
+	}
+
+	runTest := func(reporter *linux, testFilePrefix string) func(t *testing.T) {
+		return func(t *testing.T) {
+			t.Parallel()
+
+			input, err := ioutil.ReadFile(testFilePrefix + ".in")
+			if err != nil {
+				t.Fatalf("failed to read input file: %v", err)
+			}
+			output, err := ioutil.ReadFile(testFilePrefix + ".out")
+			if err != nil {
+				t.Fatalf("failed to read output file: %v", err)
+			}
+
+			result := reporter.decompileReportOpcodes(input)
+			ioutil.WriteFile(testFilePrefix+".out", result, 0640)
+
+			if !bytes.Equal(output, result) {
+				t.Fatalf("Expected:\n%v\nGot:\n%v\n", string(output), string(result))
+			}
+		}
+	}
+
+	for _, obj := range subFolders {
+		if !obj.IsDir() {
+			continue
+		}
+		reporter := prepareLinuxReporter(obj.Name(), t)
+
+		if reporter.target.BrokenCompiler != "" {
+			t.Skip("skipping the test due to broken cross-compiler:\n" + reporter.target.BrokenCompiler)
+		}
+
+		testPath := filepath.Join(archPath, obj.Name())
+		testFiles, err := ioutil.ReadDir(testPath)
+		if err != nil {
+			t.Fatalf("failed to list tests for %v: %v", obj.Name(), err)
+		}
+
+		for _, file := range testFiles {
+			if !strings.HasSuffix(file.Name(), ".in") {
+				continue
+			}
+			filePath := filepath.Join(testPath, strings.TrimSuffix(file.Name(), ".in"))
+			t.Run(obj.Name()+"/"+file.Name(), runTest(reporter, filePath))
```

Comment:
Why don't you declare runTest here inline? It looks like an unnecessary level of indirection.

Reply:
In order to decrease block nesting a bit and make the whole `TestDisassemblyInReports` look less bulky

Reply:
Then I would make it a normal function and move after this Test function. Normal functions should be considered default if closure is not really required (not capturing anything). That will have even smaller nesting level.

Reply:
In this case we're actually capturing data as we're running parallel subtests which need to get extra data. So I can move this nested function outside, but it'll still have to be the same (or nearly the same - I can call t.Run inside it).

Reply:
I would do:
```
    t.Run(obj.Name()+"/"+file.Name(), func(t *testing.T) { testDisassembly(t, reporter, testFilePrefix) })
    ...
}

func testDisassembly(t *testing.T, reporter *linux, testFilePrefix string) {
  t.Parallel()
  ... normal function goes here ...
}
```